### ISSUE TITLE
Does each worker keep a copy of the codes and implement them independly?

### ISSUE THREAD

**Issue by snsun** [horovod/horovod#1377]
**Environment:**
1. Framework: PyTorch, 
2. Framework version: 
3. Horovod version: 0.16.4
4. MPI version:
5. CUDA version: 10
6. NCCL version: 2
7. Python version: py3.6
8. OS and version:
9. GCC version:

**Your question:**
I am a little confused about how horovod works. For example, I have the following codes:

...
print("hello world")
hvd.init()
..

If I have N workers, this code will print N times. Why?  Does each worker keep a copy of the codes and implement them independently?

**Comment by tgaddair** [comment]
Hey @snsun, yes, each worker keeps its own copy of the code and runs (mostly) independently.

Horovod is based on parallel computing concepts taken from the high-performance computing community.  In that model, you run multiple copies of the same application in parallel, and then use an API like the [Message Passing Interface](https://en.wikipedia.org/wiki/Message_Passing_Interface) to synchronize state between the workers.

In Horovod, each worker process is going to train on a separate portion of the data in parallel (determined at runtime by its `rank` in MPI terms), then during the backwards pass the local gradients from each worker will be aggregated together to form a single update applied to all the workers at the same time.

Feel free to followup if you have any more questions.

**Comment by snsun** [comment]
Thank you very much. It makes sense to me. I can understand how horovod works now. 

I have another question. Does horovod support BatchNorm1d in pytorch? I have tried to add bach norm layer after lstm layer, but I found my model can not converge. I have read some answers mentioned that the mean and variance of each worker can not sync during training. Is that right?
